components/Tabs/src/MDCTabBarViewController.m

```diff
@@ -0,0 +1,224 @@
+                                         toViewController:(UIViewController *)to {
+  [self.view setNeedsLayout];
+  [self.view layoutIfNeeded];
+  from.view.hidden = YES;
+  to.view.hidden = NO;
+}
+
+// Either this has just come into existence or its array of viewControllers has changed.
+// Update the TabBar from the array of viewControllers.
+- (void)updateTabBarItems {
+  NSMutableArray<UITabBarItem *> *items = [NSMutableArray array];
+  BOOL hasTitles = NO;
+  BOOL hasImages = NO;
+  for (UIViewController *child in _viewControllers) {
+    UITabBarItem *tabBarItem = child.tabBarItem;
+    [items addObject:tabBarItem];
+    if (tabBarItem.title.length) {
+      hasTitles = YES;
+    }
+    if (tabBarItem.image) {
+      hasImages = YES;
+    }
+    [self addChildViewController:child];
+    UIView *view = child.view;
+    view.hidden = child != _selectedViewController;
+    [self.view addSubview:view];
+    [child didMoveToParentViewController:self];
+  }
+  // This class preserves the invariant that if the selected controller is not nil, it is contained
+  // in the array of viewControllers.
+  if (![_viewControllers containsObject:_selectedViewController]) {
+    self.selectedViewController = nil;
+  }
+  self.tabBar.items = items;
+  // The default height of the tab bar depends on the underlying UITabBarItems of
+  // the viewControllers.
+  if (hasImages && hasTitles) {
+    self.tabBar.itemAppearance = MDCTabBarItemAppearanceTitledImages;
+  } else if (hasImages) {
+    self.tabBar.itemAppearance = MDCTabBarItemAppearanceImages;
+  } else {
+    self.tabBar.itemAppearance = MDCTabBarItemAppearanceTitles;
+  }
+}
+
+- (nullable UIViewController *)controllerWithView:(nullable UIView*)view {
+  for (UIViewController *child in _viewControllers) {
+    if (child.view == view) {
+      return child;
+    }
+  }
+  return nil;
+}
+
+- (void)updateLayout {
+  CGRect bounds = self.view.bounds;
+  CGFloat tabBarHeight = [[_tabBar class] defaultHeightForItemAppearance:_tabBar.itemAppearance];
+  CGRect currentViewFrame = bounds;
+  CGRect tabBarFrame = CGRectMake(bounds.origin.x, bounds.origin.y + bounds.size.height,
+                                  bounds.size.width, tabBarHeight);
+  if (!_tabBarWantsToBeHidden) {
+    CGRectDivide(bounds, &tabBarFrame, &currentViewFrame, tabBarHeight, CGRectMaxYEdge);
+  }
+  _tabBar.frame = tabBarFrame;
+  _selectedViewController.view.frame = currentViewFrame;
+}
+
+#pragma mark -  MDCTabBarDelegate
+
+- (void)tabBar:(UITabBar *)tabBar didSelectItem:(UITabBarItem *)item {
+  NSUInteger index = [tabBar.items indexOfObject:item];
+  if (index < _viewControllers.count) {
+    UIViewController *newSelected = _viewControllers[index];
+    if (newSelected != self.selectedViewController) {
+      if ([_delegate respondsToSelector:@selector(tabBarController:shouldSelectViewController:)]
+          && ![_delegate tabBarController:self shouldSelectViewController:newSelected]) {
+        return;
```

Comment:
Does this fulfill the contract that returning NO prevents the tab switch? I think the tab bar will still switch and leave things inconsistent

Possibly relevant - there is a tabBar:willSelectItem: that would be a good hook for the willSelectViewController: part of this API, but no shouldSelectItem.